### PR TITLE
[WebGPU] Make the first few buffer mapping tests pass

### PR DESCRIPTION
**Pull request by litherum**
#### 2691d8a684ca5bdad0f8497c985cb4478cf0f1f4
<pre>
[WebGPU] Make the first few buffer mapping tests pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=240220">https://bugs.webkit.org/show_bug.cgi?id=240220</a>
&lt;rdar://93318423 &gt;

Reviewed by NOBODY (OOPS!).

Make getMappedRange return a ExceptionOr&lt; &gt; type. Also, code outside WebGPU.framework
needs to know how big buffers are because it needs to allocate ArrayBuffers of the correct
size. Invalid mapping ranges are represented by returning a nullptr. Also, guard against
some zero-sized copies.

Test: http/tests/webgpu/webgpu/api/operation/buffers/map.html

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
(PAL::WebGPU::BufferImpl::BufferImpl):
(PAL::WebGPU::computeRangeSize):
(PAL::WebGPU::BufferImpl::getMappedRange):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCreateImpl.cpp:
(PAL::WebGPU::create):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createBuffer):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::getMappedRange):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::copyBufferToBuffer):
(WebGPU::CommandEncoder::copyBufferToTexture):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::generateAValidationError):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::commitMTLCommandBuffer):
No new tests (OOPS!).
* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::GPUBuffer::getMappedRange):
* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map-expected.txt:
* LayoutTests/platform/ios-device-wk1/TestExpectations:
* LayoutTests/platform/ios-simulator-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
</pre>

### CODE REVIEW

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp [hunk 1]
```diff
@@ -57,13 +58,30 @@ void BufferImpl::mapAsync(MapModeFlags mapModeFlags, Size64 offset, std::optiona
     }).get());
 }
 
+static size_t computeRangeSize(uint64_t bufferSize, Size64 offset)
+{
+    auto result = checkedDifference<size_t>(bufferSize, offset);
+    if (result.hasOverflowed())
+        return 0;
```
Comment (darinadler) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:65]: I don’t understand why it’s OK to return 0 here. Seems the caller below needs to know this overflowed. Or maybe it's always safe to return something with a size of 0?

**Review by djg — APPROVED** [review]

File: Source/WebCore/Modules/WebGPU/GPUBuffer.cpp [hunk 2]
```diff
@@ -45,10 +47,14 @@ void GPUBuffer::mapAsync(GPUMapModeFlags mode, std::optional<GPUSize64> offset,
...
 
-Ref<JSC::ArrayBuffer> GPUBuffer::getMappedRange(std::optional<GPUSize64> offset, std::optional<GPUSize64> size)
+ExceptionOr<Ref<JSC::ArrayBuffer>> GPUBuffer::getMappedRange(std::optional<GPUSize64> offset, std::optional<GPUSize64> size)
 {
     auto mappedRange = m_backing->getMappedRange(offset.value_or(0), size);
-    return ArrayBuffer::create(mappedRange.source, mappedRange.byteLength);
+    if (!mappedRange.source)
+        return Exception { OperationError };
+    // FIXME: We need some kind of "detach" logic, in case someone calls destroy()
+    // on the buffer and then continues to use the ArrayBuffer.
```
Comment (darinadler) [inline Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:56]: We should make this test case ASAP. I want to know how serious the issue is.

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp [hunk 3]
```diff
@@ -57,13 +58,30 @@ void BufferImpl::mapAsync(MapModeFlags mapModeFlags, Size64 offset, std::optiona
...
-    return { pointer, static_cast<size_t>(usedSize) };
+    CheckedSize checkedOffset = offset;
+
+    if (checkedOffset.hasOverflowed())
+        return { nullptr, 0 };
+
+    CheckedSize usedSize = valueOrCompute(size, [bufferSize = m_size, offset = checkedOffset.value()]() {
+        return computeRangeSize(bufferSize, offset);
+    });
+    if (usedSize.hasOverflowed())
```
Comment (darinadler) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:79]: Maybe this check should also check for 0?

File: Source/WebGPU/WebGPU/Device.mm [hunk 4]
```diff
@@ -190,6 +190,8 @@
     }
 
     if (m_uncapturedErrorCallback) {
+        // FIXME: It's wrong to retain `this`. Instead, we should be using SharedTask.
+        // Consider someone calls setUncapturedErrorCallback() while there's a callback pending.
```
Comment (darinadler) [inline Source/WebGPU/WebGPU/Device.mm:194]: As with the other FIXME what we really need is a test for this case.

**Review by darinadler — APPROVED** [review]

**Comment by litherum** [comment]
@mwyrzykowski has taken this over.

**Pull request closed without merging**